cmd/snap-bootstrap/initramfs_install_handler.go

```diff
@@ -0,0 +1,418 @@
+func contains(s []snap.Type, e snap.Type) bool {
+	for _, a := range s {
+		if a == e {
+			return true
+		}
+	}
+	return false
+}
+
+func mountEssentials(deviceSeed seed.Seed, types []snap.Type) error {
+	// iterate over essential snaps and mount gadget and kernel
+	sd := systemd.New(systemd.SystemMode, progress.Null)
+	for _, s := range deviceSeed.EssentialSnaps() {
+		if contains(types, s.EssentialType) {
+			mountPoint := filepath.Join(boot.InitramfsRunMntDir, snapTypeToMountDir[s.EssentialType])
+			if err := doSystemdMount(s.Path, mountPoint, mountReadOnlyOptions); err != nil {
+				return err
+			}
+			activeMounts = append(activeMounts, mountInfo{
+				mountPoint: mountPoint,
+				unmountFnc: func() error {
+					return sd.Umount(mountPoint)
+				},
+				essential: true,
+			})
+		}
+	}
+	return nil
+}
+
+func umountEssentials() error {
+	for i := len(activeMounts) - 1; i >= 0; i-- {
+		if activeMounts[i].essential {
+			err := doUnmount(&activeMounts[i])
+			if err != nil {
+				return err
+			}
+			// remove from array
+			activeMounts = append(activeMounts[:i], activeMounts[i+1:]...)
+		}
+	}
+	return nil
+}
+
+func doRunInstall(deviceSeed seed.Seed, recoverySystemLabel string) error {
+	start := time.Now()
+	essentialTypes := []snap.Type{snap.TypeKernel, snap.TypeBase, snap.TypeGadget}
+	snapInfos := make(map[snap.Type]*snap.Info)
+	seedSnaps := make(map[snap.Type]*seed.Snap)
+	// collect gadget and kernel infos
+	for _, seedSnap := range deviceSeed.EssentialSnaps() {
+		typ := seedSnap.EssentialType
+		if seedSnap.Path == "" {
+			return fmt.Errorf("internal error: cannot get snap path for %s", typ)
+		}
+		snapf, err := snapfile.Open(seedSnap.Path)
+		if err != nil {
+			return fmt.Errorf("cannot open snap from %q: %v", seedSnap.Path, err)
+		}
+		snapInfo, err := snap.ReadInfoFromSnapFile(snapf, seedSnap.SideInfo)
+		if err != nil {
+			return err
+		}
+		if snapInfo.SnapType != typ {
+			return fmt.Errorf("cannot use snap info, expected %s but got %s", typ, snapInfo.SnapType)
+		}
+		seedSnaps[typ] = seedSnap
+		snapInfos[typ] = snapInfo
+	}
+	// TODO: what if we check more snaps than we need?
+	if len(snapInfos) != len(essentialTypes) {
+		return fmt.Errorf("internal error: retrieved snap infos (%d) does not match number of types (%d)", len(snapInfos), len(essentialTypes))
+	}
+	// Unset revision here actually means that the snap is local.
+	// Assign then a local revision as seeding/installing the snap would do.
+	for _, snInfo := range snapInfos {
+		if snInfo.Revision.Unset() {
+			snInfo.Revision = snap.R(-1)
+		}
+	}
+
+	// bootstrap
+	bopts := install.Options{
+		Mount: true,
+	}
+	// TODO: check function relies on model, kernel, gadget info and kernel hook
+	// kernel hook is in initrd and should check for fetures
+	// all those are ready here
+	// function should be ported from overlord/devicestate/devicemgr.go
+	// encryptionType, err := m.checkEncryption(st, deviceCtx)
+	// if err != nil {
+	// 	return err
+	// }
```

Comment:
yes, we cannot assume we will do encryption

Reply:
it's potentially next TODO to fix this